44694 hidden fields for forms

yeneastgate opened the pull request:
related to https://github.com/onOffice-Web-Org/oo-wp-plugin/issues/719
changed log:
Add hidden fields feature for forms

yeneastgate commented:
@fredericalpers Let me test it again before you change it to in-review

fredericalpers commented:
Hi @yeneastgate ,

During testing, we encountered an issue where an error occurs every time a field in the form is set to "hidden" and then the fields are saved. The following error message appears: "_An error occurred while saving the form. Please check if the form name is unique._". As a result, all the fields that have been added so far are deleted, which causes the form to be empty on the frontend.

Could you please investigate and resolve this issue?

Thank you!

![fehler1](https://github.com/onOffice-Web-Org/oo-wp-plugin/assets/88546396/7ceef210-45e4-42b5-acfb-c96aad08fcab)

yeneastgate commented:
@fredericalpers I will check it

yeneastgate commented:
@fredericalpers I apologize. 
During merging with the master branch, I missed the dbDelta function. This was the cause of the issue.

https://github.com/onOffice-Web-Org/oo-wp-plugin/assets/104892108/ff2938f1-ee47-47f4-a1bd-acfd2b63c4f5

I have now converted and fixed it. This is the video evidence. 

https://github.com/onOffice-Web-Org/oo-wp-plugin/assets/104892108/bc09cd46-a492-44d3-8e4a-2d5ace02fbc1

Could you please check and review it again?

fredericalpers commented:
Hi @yeneastgate ,

Thank you for addressing the previous issue. However, during further testing, we noticed a couple of additional problems:

- Hidden fields can still be marked as "Required." This should not be the case to prevent complications with empty fields in the frontend.
- Fields marked as hidden are not being declared as disabled in the frontend, although this was part of the requirements in [Issue 719 - Hidden fields for forms](https://github.com/onOffice-Web-Org/oo-wp-plugin/issues/719).

Could you please look into these issues and make the necessary adjustments?

Thank you for your continued effort!

yeneastgate commented:
> * Fields marked as hidden are not being declared as disabled in the frontend, although this was part of the requirements in [Issue 719 - Hidden fields for forms](https://github.com/onOffice-Web-Org/oo-wp-plugin/issues/719).

@fredericalpers  Do you mean that the fields selected as "hidden" will not be set as disabled like currently but they will be completely removed from the frontend, right?
![image](https://github.com/onOffice-Web-Org/oo-wp-plugin/assets/104892108/644136d9-6709-4434-97bc-0b269e2c8677)

fredericalpers commented:
@yeneastgate Fields marked as hidden are currently set to display: none, but they should also be disabled.

The labels of the hidden fields are still visible in the frontend. See screenshot below.

![label-still-visible](https://github.com/onOffice-Web-Org/oo-wp-plugin/assets/88546396/565759ab-8d86-4ae0-a9de-0f97b2d6c4d2)

yeneastgate commented:
@fredericalpers According to your comment, I understand that for the fields marked as "hidden", you want to remove the current "display: none" style , replace it with "disable" and show the field name as shown in the image below, is that right?
![image](https://github.com/onOffice-Web-Org/oo-wp-plugin/assets/104892108/456f01d3-24ef-4545-8871-7b8e1e9ab83a)

fredericalpers commented:
> @fredericalpers According to your comment, I understand that for the fields marked as "hidden", you want to remove the current "display: none" style , replace it with "disable" and show the field name as shown in the image below, is that right? ![image](https://private-user-images.githubusercontent.com/104892108/344135102-456f01d3-24ef-4545-8871-7b8e1e9ab83a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk1Nzc5OTYsIm5iZiI6MTcxOTU3NzY5NiwicGF0aCI6Ii8xMDQ4OTIxMDgvMzQ0MTM1MTAyLTQ1NmYwMWQzLTI0ZWYtNDU0NS04ODcxLTdiOGUxZTlhYjgzYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyOFQxMjI4MTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wZDIzOWM5NGQwYjIxNDBmNDc1NzY1MjkyYzNjYmVlNzExZjA5OTUzZGUzZDEyYTNiYWU4ZjgyYWIxOTA3YzFhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.5KA2ZvuAgtxWzpQRtkZRLcxUa7qR7Yks2QxA-Vg-4gE)

@yeneastgate  The fields should be set to display: none; and disabeld. On top of that the labels should not be visible.

yeneastgate commented:
> * Hidden fields can still be marked as "Required." This should not be the case to prevent complications with empty fields in the frontend.

I updated fields marked as "hidden" cannot be marked as "required"

> * Fields marked as hidden are not being declared as disabled in the frontend, although this was part of the requirements in [Issue 719 - Hidden fields for forms](https://github.com/onOffice-Web-Org/oo-wp-plugin/issues/719).

This feature only works for new users. For old users, please update the templates.dist folder to use this feature. You can select a template in folder templates.dist and retest because this feature is related to backward compatibility, as mentioned in the issue  https://github.com/onOffice-Web-Org/oo-wp-plugin/issues/719#issuecomment-1929160976

This is video evidence. You can check and test it again. Thank you!

https://github.com/onOffice-Web-Org/oo-wp-plugin/assets/104892108/5cd0543f-a095-4bd4-b988-3b3f7066e333

fredericalpers commented:
@yeneastgate thank you, I will review this as soon as possible :)

andernath commented on templates.dist/form/applicantform.php:
```diff
@@ -45,16 +45,20 @@
 foreach ( $pForm->getInputFields() as $input => $table ) {
```
```suggestion
foreach ( $pForm->getInputFields() as $input => $table ) {
    if($pForm->isHiddenField($input)) {
        $hiddenValues []= renderFormField($input, $pForm);
        continue;
    }
```

andernath replied:
To do not add ```<br>``` I would use an hidden value array

In line 82 and 88 I would add this array
```<?php echo implode( $hiddenValues); ?>```

andernath commented on templates.dist/form/ownerform.php:
```diff
@@ -45,22 +45,29 @@
 	foreach ( $pForm->getInputFields() as $input => $table ) {
 		$isRequired = $pForm->isRequiredField($input);
```
Please see suggestion in applicantform to remove <br> on hidden fields

andernath commented on templates.dist/form/ownerleadgeneratorform.php:
```diff
@@ -52,7 +52,9 @@
 
```
Please see suggestion in applicantform to remove <br> on hidden fields

andernath reviewed (changes requested):
@yeneastgate please take a look :)

yeneastgate commented:
@andernath I will check it asap

yeneastgate commented:
@andernath 

> Is this JS really needed?
> 
> What kind of nextSibling do you remove here? Maybe you can add a check in form template, to decide to render this sibling or not.
> 
> This seems possible to me with nextBr. See my suggestions in defaultform.php and applicantform and tell me if you agree? :)

If we use the solution you suggested, we can remove this js file. Thank you!

> I would prefer naming: $line = $label and echo without `<br>` if field is hidden. See suggestion

> I would prefer naming: $line = $label
> and echo without 'br' if field is hidden. See suggestion

I have updated it

> This can never be hidden or??

You are right, I already have the check condition in admin.js so there is no need to add code in here

> To do not add `<br>` I would use an hidden value array
> 
> In line 82 and 88 I would add this array `<?php echo implode( $hiddenValues); ?>`

Placing this code on lines 82 and 88 will cause duplicate code. See the image below
![image](https://github.com/onOffice-Web-Org/oo-wp-plugin/assets/104892108/e00aaa90-eb03-4abc-a47c-60d6f2f843a2)

I think we should put them after the "Your search criteria" rendering section.

andernath commented:
> Placing this code on lines 82 and 88 will cause duplicate code. See the image below
@yeneastgate  Ops 😁 yeah your right,.. Below search criteria is a good place 👍

andernath commented:
@yeneastgate 
For you information: I merged https://github.com/onOffice-Web-Org/oo-wp-plugin/pull/744 with database version changes into master.
You have to merge this. :)

Please tell us if this pull request is ready to review again.

By the way: I will be on vacation for the next 2 weeks. 😎

yeneastgate commented:
> @yeneastgate For you information: I merged #744 with database version changes into master. You have to merge this. :)
> 
> Please tell us if this pull request is ready to review again.
> 
> By the way: I will be on vacation for the next 2 weeks. 😎

@andernath I will check and test the branch and let you know as soon as possible. Have a great vacation! 😊

yeneastgate commented on css/onoffice-forms.css:
```diff
@@ -9,4 +9,8 @@ span.onoffice-pleasefill {
+}
+
+.oo-form .hidden-field {
+	display: none;
```
@andernath  We should keep the parent class "oo-form" because some themes will have their own CSS styles to customize input fields. If you don't use the parent class, the priority of the above style snippet will be lower, and then the attribute "display:none" will not be executed.

pospisk commented on templates.dist/fields.php:
```diff
@@ -122,6 +122,7 @@ function renderFormField(string $fieldName, onOffice\WPlugin\Form $pForm, bool $
 		$selectedValue = $pForm->getFieldValue($fieldName, true);
 		$isRangeValue = $pForm->isSearchcriteriaField($fieldName) && $searchCriteriaRange;
 		$fieldLabel = $pForm->getFieldLabel($fieldName, true);
+		$isHiddenField = $pForm->isHiddenField($fieldName);
```
Please define the hidden attribute here 
using:
```
$isHiddenField = $pForm->isHiddenField($fieldName);
$hiddenAttribute = $isHiddenField ? 'class="oo-hidden-field" readonly ' : '';
```

instead of:
```
if ($isHiddenField === true) {
	$inputType .= 'class="oo-hidden-field" disabled ';
}
```

pospisk commented on templates.dist/fields.php:
```diff
@@ -172,14 +175,16 @@ function renderFormField(string $fieldName, onOffice\WPlugin\Form $pForm, bool $
 			$value = 'value="' . esc_attr($pForm->getFieldValue($fieldName, true)) . '"';
-
-			if ($typeCurrentInput == onOffice\WPlugin\Types\FieldTypes::FIELD_TYPE_BOOLEAN) {
+			if ($isHiddenField === true) {
+				$inputType .= 'class="oo-hidden-field" disabled ';
```
Please remove:
```
if ($isHiddenField === true) {
	$inputType .= 'class="oo-hidden-field" disabled ';
} else
```

to restore the original if statement:
```
if ($typeCurrentInput == onOffice\WPlugin\Types\FieldTypes::FIELD_TYPE_BOOLEAN) {
```

afterwards to make it work please insert the following `$inputType .= $hiddenAttribute;` around line 210 

```
} elseif (
	$typeCurrentInput === FieldTypes::FIELD_TYPE_DATETIME
) {
	$inputType = 'type="datetime-local" step="1" ';
}

// here

if (
	$isRangeValue && $pForm->inRangeSearchcriteriaInfos($fieldName) &&
	count($pForm->getSearchcriteriaRangeInfosForField($fieldName)) > 0
) {
```

Bug:
Currently when a checkbox is set to hidden the input type="" is stuck on "text" when it should be "checkbox". This causes an incorrect value in the field and they aren't submitted in the form. 

<img width="1149" alt="hidden" src="https://github.com/user-attachments/assets/82952c8d-66ad-4b5e-a7c0-181272703647">

pospisk commented on templates.dist/fields.php:
```diff
@@ -214,7 +219,8 @@ function renderFormField(string $fieldName, onOffice\WPlugin\Form $pForm, bool $
 				}
 			} elseif ($typeCurrentInput === FieldTypes::FIELD_TYPE_DATATYPE_TINYINT) {
-				$output = '<fieldset>
+				$additionHidden = $isHiddenField === true ? 'class="oo-hidden-field" disabled ' : '';
+				$output = '<fieldset '.$additionHidden.'>
```
Please remove `$additionHidden`, you can use `$hiddenAttribute` instead which is defined at line 126.

pospisk commented on templates.dist/form/defaultform.php:
```diff
@@ -65,16 +65,22 @@
+				echo $additionMessage . ':<br>';
+				echo '<textarea name="message">' . $pForm->getFieldValue('message') . '</textarea><br>';
+			} else {
+				echo '<textarea name="message" class="oo-hidden-field" disabled">' . $pForm->getFieldValue('message') . '</textarea>';
```
There is a stray `"` double-quotation after `disabled` and please use `readonly` instead of `disabled`.

pospisk commented on templates.dist/form/ownerform.php:
```diff
@@ -58,9 +63,14 @@
+				$line .= $additionMessage . ':<br>';
+				$line .= '<textarea name="message">' . $pForm->getFieldValue('message') . '</textarea><br>';
+			} else {
+				$line = '<textarea name="message" class="oo-hidden-field" disabled">'.$pForm->getFieldValue('message').'</textarea><br>';
```
There is a stray `"` double-quotation after `disabled` and please use `readonly` instead of `disabled`.

pospisk reviewed (changes requested):
@yeneastgate 
Please take a look at the requested changes.

yeneastgate commented:
@pospisk I have refactored the code, and after checking, I found that in some cases, such as select elements, the readonly attribute cannot be applied. Therefore, I have added the disabled attribute for those instances. You can see this in the video.

https://github.com/user-attachments/assets/26826467-b67a-4128-acd8-6f314a46e17b


Please review the changes and let me know your opinion.

pospisk commented on plugin/FormPost.php:
```diff
@@ -227,6 +231,42 @@ private function buildFormData(DataFormConfiguration $pFormConfig, $formNo): For
+			if (count($pDefaultFields)) $values = array_merge($values, $pDefaultFields);
+		}
+
+		array_walk($values, function($value, $key) {
+			if ($value === "0" || !empty($value)) {
+				$_POST[$key] = $value;
+			}
+		});
```
<img width="1371" alt="form" src="https://github.com/user-attachments/assets/875903f0-16f9-4fc5-beaa-f991d4721033">

**Issues:** 
- If a field has a default value set, we can remove it from the HTML but it will still be submitted.
- Changing the value in HTML can’t overwrite the default value set.
- Hidden checkboxes with a default value "y" submit "1" which is interpreted as "n".

**Fix:**
_plugin/FormPost.php_

```
private function readDefaultValueForHiddenFieldsByFormId(int $formId, array $hiddenFields)
...
array_walk($values, function($value, $key) {
	if ($value === "0" || !empty($value) && !isset($_POST[$key])) {
		$_POST[$key] = $value;
	}
});
```
We need `!isset($_POST[$key])` to make sure we can overwrite the default value set in the plugin

pospisk reviewed (changes requested):
The fix within FormPost.php requires changes in the fields.php and not all parts of the file are visible within the review. (see next message)

yeneastgate commented:
@pospisk I will check it.

yeneastgate commented:
@pospisk Based on the issues and solutions you provided, it seems that we can still edit the HTML by using the browser's developer tools or "View Source". However, according to the initial requirements, you want us to disable this behavior. https://github.com/onOffice-Web-Org/oo-wp-plugin/issues/719#issue-2074416434
> While the fields values that are marked with the hidden attribute are not displayed to the user in the page's content, it is visible and can be edited using any browser's developer tools or "View Source" functionality. Therefore we also need to add the "disabled" attribute to all fields that are marked as hidden.

Please confirm the following:
- Editing HTML using the browser's developer tools or "View Source".
- If a field is marked as hidden, its type should be changed to hidden.

yeneastgate commented:
@pospisk Based on the issues and solutions you provided, it seems that we can still edit the HTML by using the browser's developer tools or "View Source". However, according to the initial requirements, you want us to disable this behavior. https://github.com/onOffice-Web-Org/oo-wp-plugin/issues/719#issue-2074416434
> While the fields values that are marked with the hidden attribute are not displayed to the user in the page's content, it is visible and can be edited using any browser's developer tools or "View Source" functionality. Therefore we also need to add the "disabled" attribute to all fields that are marked as hidden.

Please confirm the following:
- Can we edit the HTML using the browser's developer tools or "View Source" and have the modified values submitted instead of the default values?
- If a field is marked as hidden, do you want all types to be changed to hidden?

fredericalpers commented:
> @pospisk Based on the issues and solutions you provided, it seems that we can still edit the HTML by using the browser's developer tools or "View Source". However, according to the initial requirements, you want us to disable this behavior. [#719 (comment)](https://github.com/onOffice-Web-Org/oo-wp-plugin/issues/719#issue-2074416434)
> 
> > While the fields values that are marked with the hidden attribute are not displayed to the user in the page's content, it is visible and can be edited using any browser's developer tools or "View Source" functionality. Therefore we also need to add the "disabled" attribute to all fields that are marked as hidden.
> 
> Please confirm the following:
> 
> * Can we edit the HTML using the browser's developer tools or "View Source" and have the modified values submitted instead of the default values?
> * If a field is marked as hidden, do you want all types to be changed to hidden?

Please go ahead and implement @pospisk suggested solution. 

> * Can we edit the HTML using the browser's developer tools or "View Source" and have the modified values submitted instead of the default values?
> * If a field is marked as hidden, do you want all types to be changed to hidden?

I can confirm both :)

yeneastgate commented:
@fredericalpers  @pospisk I have updated the code and here is evidence https://files.fm/f/sdeeqnrw9h
Please review and recheck it. Thanks!

pospisk reviewed (approved):
Excellent, thank you very much.
